### PR TITLE
RA-1653: Program Status widgets, single call to location REST endpoint.

### PR DESCRIPTION
**Pull request by Ruhanga**
Ticket : https://issues.openmrs.org/browse/RA-1653

### CODE REVIEW

File: omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js [hunk 1]
```diff
@@ -197,12 +197,19 @@ export default class ProgramStatusController {
...
-            this.programLocations = response.results;
-        })
+        var locationsPromise = this.$window.sessionContext.locationsPromise
+        if(!locationsPromise) {
+            locationsPromise = this.openmrsRest.get("location", {
+                v: "custom:display,uuid",
+                tag: this.config.locationTag
+            })
+            // make the location promise available for other program widgets
+            this.$window.sessionContext.locationsPromise = locationsPromise
```
Comment (Ruhanga) [inline omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js]: Hi @mogoodrich, I ended up using the `sessionContext` object to contain the the locations promise so all the existing program status widgets can use it once to populate their locations. Do you anticipate any issue with this approach? Thanks.

Comment (mogoodrich) [reply omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js]: Thanks @Ruhanga .  I'm not familiar with window.sessionContext... can you point me to do some documentation of it?  Another option would be to use window.sessionStorage... https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage

Comment (Ruhanga) [reply omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js]: Ahh, great. Thanks @mogoodrich. `sessionContext` is defined [in appui](https://github.com/openmrs/openmrs-module-appui/blob/master/omod/src/main/webapp/fragments/decorator/standardEmrPage.gsp#L28) and is used mainly for the locale info. I'm going to use this `window.sessionStorage` object instead. Thank you :-)

Comment (mogoodrich) [reply omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js]: Oh interesting!  It doesn't look like the "window.sessionContext" is used often (or at all) so sessionStorage is probably a better bet.

Comment (Ruhanga) [reply omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js]: @mogoodrich, so it is impossible to store objects with functions in the `sessionStorage`. `sessionStorage` does not handle events. This makes it difficult to store a promise or use some event handling mechanism that updates multiple program status widgets with the fetched locations. Perhaps I may create another `window.` say `programEnrollmentLocationsPromise`.  `programEnrollmentLocationsPromise` is quite verbose though. :(

Comment (mogoodrich) [reply omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js]: @Ruhanga ah... feel free to use sessionContext then, I'd just add a comment to the code explaining where/how that is defined.

**Comment by Ruhanga** [comment]
Great. Thanks @mogoodrich.

**Pull request merged**